x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java

```diff
@@ -148,10 +164,75 @@ public static TemporalAmount parseTemporalAmout(Number value, String qualifier,
         };
     }
 
+    /**
+     * Converts a string or date expression to long.
+     */
+    private static Long foldExpressionToLong(Expression exp) {
+        Object value = Foldables.valueOf(exp);
+        if (DataTypes.isDateTime(exp.dataType())) {
+            return ((Number) value).longValue();
+        } else if (DataTypes.isString(exp.dataType())) {
+            return convertDatetimeStringToLong(((BytesRef) value).utf8ToString());
+        }
+        return null;
+    }
+
+    /**
+     * Converts a string expression to ChronoField.
+     */
+    private static ChronoField foldExpressionToChronoField(Expression exp) {
+        Object value = Foldables.valueOf(exp);
+        ChronoField chronoField = null;
+        if (exp.dataType() == DataTypes.KEYWORD) {
+            try {
+                BytesRef br = BytesRefs.toBytesRef(value);
+                chronoField = ChronoField.valueOf(br.utf8ToString().toUpperCase(Locale.ROOT));
+            } catch (Exception e) {
+                return null;
+            }
+        }
+        return chronoField;
+    }
+
+    private static BytesRef convertStringToIP(BytesRef value) {
+        return parseIP(value.utf8ToString());
+    }
+
+    /**
+     * Returns Datetime in long according to a ChronoField provided.
+     */
+    public static long extractLongChronoField(long value, Object chronoField, ZoneId zone) {
+        ChronoField chrono = null;
+        if (chronoField instanceof BytesRef cf) {
+            chrono = ChronoField.valueOf(cf.utf8ToString().toUpperCase(Locale.ROOT));
+        } else if (chronoField instanceof ChronoField cf) {
+            chrono = cf;
+        } else {
+            throw new IllegalArgumentException("unsupported type [" + chrono.getClass() + "]");
```

Comment:
Why the exception ? Shouldn't it be null like the other methods?

Reply:
Will move the exception to caller.